active/0000-manually-drop.md

```diff
@@ -0,0 +1,194 @@
+// v.inline = [uninit, uninit, ... ]
+```
+
+When a `SmallVec` with the above definition is dropped, the
+destructors of all 8 `T`s in `inline` will always be run (no matter
+what happens in the `Drop` implementation of `SmallVec`), leading to
+dramatic unsafety: in the first and last cases, destructors would be
+running on uninitialised data, and in the second, it would correctly
+destroy the first element (`"foo"`) but the last 7 are uninitialised.
+
+With the `ManuallyDrop` type given in this RFC, `SmallVec` could be
+defined as
+
+```rust
+// (sketch, without `unsafe {}`s/casts etc. for clarity)
+
+struct SmallVec<T> {
+    length: uint,
+    capacity: uint,
+    pointer: *mut T,
+
+    // now these 8 T's are not automatically destroyed when a SmallVec
+    // goes out of scope.
+    inline: ManuallyDrop<[T, .. 8]>
+}
+
+impl<T> Drop for SmallVec<T> {
+    fn drop(&mut self) {
+        if !self.pointer.is_null() { // heap allocated
+            // same as Vec<T>'s drop
+            for i in range(0, self.length) {
+                ptr::read(self.pointer.offset(i));
+            }
+            alloc::heap::deallocate(self.pointer, ...);
+        } else {
+            // run destructors on the first `self.length` elements of
+            // `self.inline`, but no others.
+            for i in range(0, self.length) {
+                ptr::read(&self.inline.get().offset(i));
+            }
+        }
+    }
+}
+```
+
+This definition is now safe: destructors run on exactly those elements
+that are valid and no others.
+
+
+To be precise, at the time of writing, Rust defines that destructors
+are safe to run on zeroed values (and guarantees it via the so called
+"drop flag"), meaning this can currently be made safe via:
+
+```rust
+impl<T> Drop for SmallVec<T> {
+    fn drop(&mut self) {
+        // same as above.
+
+        // zero the whole array: it's fine for the `inner` field to be
+        // redestroyed now.
+        mem::set_memory(&mut self.inner as *mut T, 0, self.inner.len());
+    }
+}
+```
+
+However, the drop flag is likely to disappear:
+[#5016](https://github.com/mozilla/rust/issues/5016).
+
+# Detailed design
+
+A new lang-item type equivalent to
+
+```rust
+pub struct ManuallyDrop<T> {
+    pub data: T
+}
+```
+
+would be provided, and the compiler would know to *not* run the
+destructor of the `T` when a `ManuallyDrop<T>` instance goes out of
+scope. That is,
+
+```rust
+struct X { num: int };
+impl Drop for X {
+    fn drop(&mut self) { println!("{}", self.num) }
+}
+
+fn main() {
+    let normal = X { x: 0 };
+    let new = ManuallyDrop { data: X { x: 1 } };
+}
+```
+
+would print only `0`, *not* `1`.
+
+This would be modelled after the `UnsafeCell` type, providing similar
+methods:
+
+- `fn get(&self) -> *const T`
+- `fn get_mut(&mut self) -> *mut T`
+- `unsafe fn unwrap(self) -> T`
+
+The `data` field is public to allow for static initialisation (again,
+similar to `UnsafeCell`).
+
+# Drawbacks
+
+This is quite dangerous.
```

Comment:
Dangerous as in how? Dangerous to use (we have `unsafe` exactly for that and leaking is not unsafe anymore either)? Dangerous to implement and commit to this (why?)?